storage/innobase/buf/buf0lru.cc

```diff
@@ -986,16 +986,16 @@ buf_LRU_block_free_non_file_page(
 	ut_ad(!block->page.in_LRU_list);
 	ut_ad(!block->page.hash);
 
-	block->page.set_state(buf_page_t::NOT_USED);
-
-	MEM_UNDEFINED(block->page.frame, srv_page_size);
 	/* Wipe page_no and space_id */
 	static_assert(FIL_PAGE_OFFSET % 4 == 0, "alignment");
 	memset_aligned<4>(block->page.frame + FIL_PAGE_OFFSET, 0xfe, 4);
 	static_assert(FIL_PAGE_ARCH_LOG_NO_OR_SPACE_ID % 4 == 2,
 		      "not perfect alignment");
 	memset_aligned<2>(block->page.frame + FIL_PAGE_ARCH_LOG_NO_OR_SPACE_ID,
 			  0xfe, 4);
```

Comment:
If I remember correctly, this code is here only to make it more likely to catch any misuse of the buffer pool. This code can be replaced with an explicit call to a new member function `block->page.set_unused()`, which would inform the operating system that the page frame is no longer needed.

Reply:
Any blocks that are in the `buf_pool.free` list should be advertised as free to the operating system. This function `buf_LRU_block_free_non_file_page()` is freeing blocks to that list, and `buf_LRU_get_free_only()` is allocating blocks from `buf_pool.free`. Your revised commit is modifying both. It is correct to call `block->page.set_os_unused();` right after these writes. But, it should also be correct to remove those writes in the first place, like I suggested. At the end of `buf_LRU_block_free_non_file_page()` we have a better safety mechanism to catch unwanted access to the page frame: the `MEM_NOACCESS` call "poisons" the page frame in AddressSanitizer and Valgrind.

I see that `buf_block_t::initialise()` will be immediately followed by some code that will use the page frame. Your revised patch changes that function only, not `buf_page_t::init()`. Also that looks correct to me.